File: py/poetrydb_sdk.py
# Poetrydb SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import PoetrydbUtility
from core.spec import PoetrydbSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import PoetrydbBaseFeature
from features import _make_feature


class PoetrydbSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = PoetrydbUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return PoetrydbUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = PoetrydbSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def author(self):
        """Idiomatic facade: client.author.list() / client.author.load({"id": ...})."""
        from entity.author_entity import AuthorEntity
        cached = getattr(self, "_author", None)
        if cached is None:
            cached = AuthorEntity(self, None)
            self._author = cached
        return cached

    def Author(self, data=None):
        # Deprecated: use client.author instead.
        from entity.author_entity import AuthorEntity
        return AuthorEntity(self, data)


    @property
    def authorab(self):
        """Idiomatic facade: client.authorab.list() / client.authorab.load({"id": ...})."""
        from entity.authorab_entity import AuthorabEntity
        cached = getattr(self, "_authorab", None)
        if cached is None:
            cached = AuthorabEntity(self, None)
            self._authorab = cached
        return cached

    def Authorab(self, data=None):
        # Deprecated: use client.authorab instead.
        from entity.authorab_entity import AuthorabEntity
        return AuthorabEntity(self, data)


    @property
    def combined_search(self):
        """Idiomatic facade: client.combined_search.list() / client.combined_search.load({"id": ...})."""
        from entity.combined_search_entity import CombinedSearchEntity
        cached = getattr(self, "_combined_search", None)
        if cached is None:
            cached = CombinedSearchEntity(self, None)
            self._combined_search = cached
        return cached

    def CombinedSearch(self, data=None):
        # Deprecated: use client.combined_search instead.
        from entity.combined_search_entity import CombinedSearchEntity
        return CombinedSearchEntity(self, data)


    @property
    def combined_search_with_field(self):
        """Idiomatic facade: client.combined_search_with_field.list() / client.combined_search_with_field.load({"id": ...})."""
        from entity.combined_search_with_field_entity import CombinedSearchWithFieldEntity
        cached = getattr(self, "_combined_search_with_field", None)
        if cached is None:
            cached = CombinedSearchWithFieldEntity(self, None)
            self._combined_search_with_field = cached
        return cached

    def CombinedSearchWithField(self, data=None):
        # Deprecated: use client.combined_search_with_field instead.
        from entity.combined_search_with_field_entity import CombinedSearchWithFieldEntity
        return CombinedSearchWithFieldEntity(self, data)


    @property
    def line(self):
        """Idiomatic facade: client.line.list() / client.line.load({"id": ...})."""
        from entity.line_entity import LineEntity
        cached = getattr(self, "_line", None)
        if cached is None:
            cached = LineEntity(self, None)
            self._line = cached
        return cached

    def Line(self, data=None):
        # Deprecated: use client.line instead.
        from entity.line_entity import LineEntity
        return LineEntity(self, data)


    @property
    def linecount(self):
        """Idiomatic facade: client.linecount.list() / client.linecount.load({"id": ...})."""
        from entity.linecount_entity import LinecountEntity
        cached = getattr(self, "_linecount", None)
        if cached is None:
            cached = LinecountEntity(self, None)
            self._linecount = cached
        return cached

    def Linecount(self, data=None):
        # Deprecated: use client.linecount instead.
        from entity.linecount_entity import LinecountEntity
        return LinecountEntity(self, data)


    @property
    def poemcount(self):
        """Idiomatic facade: client.poemcount.list() / client.poemcount.load({"id": ...})."""
        from entity.poemcount_entity import PoemcountEntity
        cached = getattr(self, "_poemcount", None)
        if cached is None:
            cached = PoemcountEntity(self, None)
            self._poemcount = cached
        return cached

    def Poemcount(self, data=None):
        # Deprecated: use client.poemcount instead.
        from entity.poemcount_entity import PoemcountEntity
        return PoemcountEntity(self, data)


    @property
    def random(self):
        """Idiomatic facade: client.random.list() / client.random.load({"id": ...})."""
        from entity.random_entity import RandomEntity
        cached = getattr(self, "_random", None)
        if cached is None:
            cached = RandomEntity(self, None)
            self._random = cached
        return cached

    def Random(self, data=None):
        # Deprecated: use client.random instead.
        from entity.random_entity import RandomEntity
        return RandomEntity(self, data)


    @property
    def title(self):
        """Idiomatic facade: client.title.list() / client.title.load({"id": ...})."""
        from entity.title_entity import TitleEntity
        cached = getattr(self, "_title", None)
        if cached is None:
            cached = TitleEntity(self, None)
            self._title = cached
        return cached

    def Title(self, data=None):
        # Deprecated: use client.title instead.
        from entity.title_entity import TitleEntity
        return TitleEntity(self, data)


    @property
    def titleab(self):
        """Idiomatic facade: client.titleab.list() / client.titleab.load({"id": ...})."""
        from entity.titleab_entity import TitleabEntity
        cached = getattr(self, "_titleab", None)
        if cached is None:
            cached = TitleabEntity(self, None)
            self._titleab = cached
        return cached

    def Titleab(self, data=None):
        # Deprecated: use client.titleab instead.
        from entity.titleab_entity import TitleabEntity
        return TitleabEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk

File: rb/Poetrydb_sdk.rb
# Poetrydb SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'Poetrydb_types'


class PoetrydbSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = PoetrydbUtility.new
    @_utility = utility

    config = PoetrydbConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = PoetrydbHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = PoetrydbHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, PoetrydbFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    PoetrydbUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = PoetrydbHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = PoetrydbHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = PoetrydbHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = PoetrydbSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    utility.make_fetch_def.call(ctx)
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue PoetrydbError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = PoetrydbHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = PoetrydbHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Idiomatic facade: client.author.list / client.author.load({ "id" => ... })
  def author
    require_relative 'entity/author_entity'
    @author ||= AuthorEntity.new(self, nil)
  end

  # Deprecated: use client.author instead.
  def Author(data = nil)
    require_relative 'entity/author_entity'
    AuthorEntity.new(self, data)
  end


  # Idiomatic facade: client.authorab.list / client.authorab.load({ "id" => ... })
  def authorab
    require_relative 'entity/authorab_entity'
    @authorab ||= AuthorabEntity.new(self, nil)
  end

  # Deprecated: use client.authorab instead.
  def Authorab(data = nil)
    require_relative 'entity/authorab_entity'
    AuthorabEntity.new(self, data)
  end


  # Idiomatic facade: client.combined_search.list / client.combined_search.load({ "id" => ... })
  def combined_search
    require_relative 'entity/combined_search_entity'
    @combined_search ||= CombinedSearchEntity.new(self, nil)
  end

  # Deprecated: use client.combined_search instead.
  def CombinedSearch(data = nil)
    require_relative 'entity/combined_search_entity'
    CombinedSearchEntity.new(self, data)
  end


  # Idiomatic facade: client.combined_search_with_field.list / client.combined_search_with_field.load({ "id" => ... })
  def combined_search_with_field
    require_relative 'entity/combined_search_with_field_entity'
    @combined_search_with_field ||= CombinedSearchWithFieldEntity.new(self, nil)
  end

  # Deprecated: use client.combined_search_with_field instead.
  def CombinedSearchWithField(data = nil)
    require_relative 'entity/combined_search_with_field_entity'
    CombinedSearchWithFieldEntity.new(self, data)
  end


  # Idiomatic facade: client.line.list / client.line.load({ "id" => ... })
  def line
    require_relative 'entity/line_entity'
    @line ||= LineEntity.new(self, nil)
  end

  # Deprecated: use client.line instead.
  def Line(data = nil)
    require_relative 'entity/line_entity'
    LineEntity.new(self, data)
  end


  # Idiomatic facade: client.linecount.list / client.linecount.load({ "id" => ... })
  def linecount
    require_relative 'entity/linecount_entity'
    @linecount ||= LinecountEntity.new(self, nil)
  end

  # Deprecated: use client.linecount instead.
  def Linecount(data = nil)
    require_relative 'entity/linecount_entity'
    LinecountEntity.new(self, data)
  end


  # Idiomatic facade: client.poemcount.list / client.poemcount.load({ "id" => ... })
  def poemcount
    require_relative 'entity/poemcount_entity'
    @poemcount ||= PoemcountEntity.new(self, nil)
  end

  # Deprecated: use client.poemcount instead.
  def Poemcount(data = nil)
    require_relative 'entity/poemcount_entity'
    PoemcountEntity.new(self, data)
  end


  # Idiomatic facade: client.random.list / client.random.load({ "id" => ... })
  def random
    require_relative 'entity/random_entity'
    @random ||= RandomEntity.new(self, nil)
  end

  # Deprecated: use client.random instead.
  def Random(data = nil)
    require_relative 'entity/random_entity'
    RandomEntity.new(self, data)
  end


  # Idiomatic facade: client.title.list / client.title.load({ "id" => ... })
  def title
    require_relative 'entity/title_entity'
    @title ||= TitleEntity.new(self, nil)
  end

  # Deprecated: use client.title instead.
  def Title(data = nil)
    require_relative 'entity/title_entity'
    TitleEntity.new(self, data)
  end


  # Idiomatic facade: client.titleab.list / client.titleab.load({ "id" => ... })
  def titleab
    require_relative 'entity/titleab_entity'
    @titleab ||= TitleabEntity.new(self, nil)
  end

  # Deprecated: use client.titleab instead.
  def Titleab(data = nil)
    require_relative 'entity/titleab_entity'
    TitleabEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = PoetrydbSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end
